frontend/src/metabase/public/containers/PublicQuestion/PublicQuestion.tsx

```diff
@@ -0,0 +1,238 @@
+      dispatch(setErrorPage(error));
+    }
+  });
+
+  const setParameterValue = async (parameterId: ParameterId, value: any) => {
+    setParameterValues(prevParameterValues => ({
+      ...prevParameterValues,
+      [parameterId]: value,
+    }));
+  };
+
+  const setParameterValueToDefault = (parameterId: ParameterId) => {
+    const parameters = getParameters();
+    const parameter = parameters.find(({ id }) => id === parameterId);
+    if (parameter) {
+      setParameterValue(parameterId, parameter.default);
+    }
+  };
+
+  const run = useCallback(async () => {
+    if (!card) {
+      return;
+    }
+
+    const parameters = card.parameters || getParametersFromCard(card);
+
+    try {
+      setResult(null);
+
+      let newResult;
+      if (token) {
+        // embeds apply parameter values server-side
+        newResult = await maybeUsePivotEndpoint(
+          EmbedApi.cardQuery,
+          card,
+        )({
+          token,
+          ...getParameterValuesBySlug(parameters, parameterValues),
+        });
+      } else if (uuid) {
+        // public links currently apply parameters client-side
+        const datasetQuery = applyParameters(card, parameters, parameterValues);
+        newResult = await maybeUsePivotEndpoint(
+          PublicApi.cardQuery,
+          card,
+        )({
+          uuid,
+          parameters: JSON.stringify(datasetQuery.parameters),
+        });
+      } else {
+        throw { status: 404 };
+      }
+
+      setResult(newResult);
+    } catch (error) {
+      console.error("error", error);
+      dispatch(setErrorPage(error));
+    }
+  }, [card, dispatch, parameterValues, token, uuid]);
+
+  useEffect(() => {
+    run();
+  }, [run]);
+
+  const getParameters = () => {
+    if (!card) {
+      return [];
+    }
+
+    return getCardUiParameters(
+      card,
+      metadata,
+      {},
+      card.parameters || undefined,
+    );
+  };
+
+  const question = new Question(card, metadata);
+
+  const actionButtons = result && (
+    <QueryDownloadWidget
+      className={cx(CS.m1, CS.textMediumHover)}
+      question={question}
+      result={result}
+      uuid={uuid}
+      token={token}
+    />
+  );
+
+  const { bordered, hide_download_button, hide_parameters, theme, titled } =
+    useEmbedFrameOptions({ location });
+
+  return (
+    <EmbedFrame
+      name={card && card.name}
+      description={card && card.description}
+      actionButtons={actionButtons}
+      question={question}
+      parameters={getParameters()}
+      parameterValues={parameterValues}
+      setParameterValue={setParameterValue}
+      enableParameterRequiredBehavior
+      setParameterValueToDefault={setParameterValueToDefault}
+      bordered={bordered}
+      hide_download_button={hide_download_button}
+      hide_parameters={hide_parameters}
+      theme={theme}
+      titled={titled}
+    >
+      <LoadingAndErrorWrapper
+        className={CS.flexFull}
+        loading={!result}
```

Comment:
Removing `|| !initialized` here makes sense since `result` will become available right after we set `initialized` if anything would change, the loading would finish one tick faster than before.